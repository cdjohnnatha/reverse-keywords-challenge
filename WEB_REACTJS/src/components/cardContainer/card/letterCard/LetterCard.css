.card-class {
  box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
  transition: 0.3s;
  border-radius: 5px;
  padding: 1em;
  border: 1px solid lightblue;
  cursor: pointer;
}

.card-class:hover {
  background-color: grey
}

.flex-container {
  display: flex;
  justify-content: center;
}

.box-margin {
  margin: 1em 0.2em 1em 0.2em;
}

.green-bg {
  background-color: lightgreen;
}

.red-bg {
  background-color: lightcoral;
}