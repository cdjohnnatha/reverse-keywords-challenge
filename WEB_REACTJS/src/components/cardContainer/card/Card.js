import React from 'react';
import { string, func, number, bool } from 'prop-types';
import LetterCard from './letterCard/LetterCard';
import './Card.css';

const Card = ({
  typedWords,
  onSingleLetterClickHandler,
  cardHistoryIndex,
  isFromWordHistory,
}) => {
  let lettersArr = null;
  let renderLetters = null;
  if (typedWords && typedWords.length > 0) {
    lettersArr = typedWords.split('');
    renderLetters = lettersArr.map((letter, index) => (
      <LetterCard
        key={`card-letter-${letter}-${index}`}
        letter={letter}
        onSingleLetterClickHandler={onSingleLetterClickHandler}
        letterIndexPosition={index}
        isFromWordHistory={isFromWordHistory}
        cardHistoryIndex={cardHistoryIndex}
      />
    ));
  }

  return <div className="container">{renderLetters}</div>;
};

Card.propTypes = {
  typedWords: string,
  onSingleLetterClickHandler: func.isRequired,
  cardHistoryIndex: number,
  isFromWordHistory: bool,
};

Card.defaultProps = {
  typedWords: 'Kenna',
  isFromWordHistory: false,
  cardHistoryIndex: null,
};

export default Card;
