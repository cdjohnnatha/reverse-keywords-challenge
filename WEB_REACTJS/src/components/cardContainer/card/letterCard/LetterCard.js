import React from 'react';
import { string, func, number, bool } from 'prop-types';
import './LetterCard.css';

const LetterCard = ({
  letter,
  letterIndexPosition,
  onSingleLetterClickHandler,
  isFromWordHistory,
  cardHistoryIndex,
}) => {
  let classList = 'card-class flex-container box-margin';
  if (isFromWordHistory) classList += ' red-bg';
  else classList += ' green-bg';

  return (
    <div
      className={classList}
      onClick={() =>
        onSingleLetterClickHandler(letterIndexPosition, cardHistoryIndex)
      }
    >
      {letter}
    </div>
  );
};

LetterCard.propTypes = {
  letter: string.isRequired,
  onSingleLetterClickHandler: func.isRequired,
  letterIndexPosition: number.isRequired,
  cardHistoryIndex: number,
  isFromWordHistory: bool.isRequired,
};

LetterCard.defaultProps = {
  cardHistoryIndex: null,
};

export default LetterCard;
