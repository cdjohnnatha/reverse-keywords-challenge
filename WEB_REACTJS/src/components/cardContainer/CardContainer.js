import React, { useState } from 'react';
import Card from './card/Card';
import InputField from './inputField/InputField';
import { removeArrayElement } from '../../helpers/array-helpers';
import SendReverseWordButton from './sendReverseWordButton/SendReverseWordButton';
import ErrorMessage from '../errorMessage/ErrorMessage';

const CardContainer = () => {
  const [typedWords, setTypedWords] = useState({
    recentInput: 'Kenna',
    historyWords: [],
  });

  const [errorMessage, setErrorMessage] = useState();
  let renderWordsHistory = null;

  /**
   * @function onInputFieldChangeHandler
   * @description It will be used to change the input field state and value.
   * @param {Object} event - input event object.
   * @param {Object} event.target - input event target.
   * @param {Object} event.target.value - value typed which it will be used on typedWords state.
   */
  const onInputFieldChangeHandler = ({ target }) => {
    typedWords.recentInput = target.value;
    setTypedWords({ ...typedWords });
  };

  /**
   * @function onSingleLetterCardClickHandler
   * @description It will be used to remove a letter from a word. If the word is considered as
   * a recenpt input then it will be removed from it and update the state for typedWords,
   * if a letter position be considered from typedWords.historyWords then it will retrieve
   * the word from array, remove the right letter and the update the state.
   * @param {Number} letterIndexPosition - Letter position from typedWords.recentInput
   * to be removed.
   * @param {Number|undefined} cardHistoryIndex - This number it will show up when the word belongs
   * to a history and it will identify the history word position to.
   */
  const onSingleLetterCardClickHandler = (
    letterIndexPosition,
    cardHistoryIndex
  ) => {
    let wordToBeLetterRemoved = typedWords.recentInput;
    const isLetterFromWordHistory = typeof cardHistoryIndex === 'number';
    if (isLetterFromWordHistory) {
      wordToBeLetterRemoved = typedWords.historyWords[cardHistoryIndex];
    }
    const newWordAfterRemoval = removeArrayElement(
      wordToBeLetterRemoved,
      letterIndexPosition
    ).join('');

    if (isLetterFromWordHistory) {
      typedWords.historyWords[cardHistoryIndex] = newWordAfterRemoval;
      if (typedWords.historyWords[cardHistoryIndex].length === 0) {
        typedWords.historyWords = removeArrayElement(
          typedWords.historyWords,
          cardHistoryIndex
        );
      }
    } else typedWords.recentInput = newWordAfterRemoval;
    setTypedWords({ ...typedWords });
  };

  if (typedWords.historyWords.length > 0) {
    renderWordsHistory = typedWords.historyWords.map((wordHistory, index) => (
      <Card
        key={`word-history-key-${index}`}
        typedWords={wordHistory}
        onSingleLetterClickHandler={onSingleLetterCardClickHandler}
        cardHistoryIndex={index}
        isFromWordHistory
      />
    ));
    renderWordsHistory = <div id="card-history">{renderWordsHistory}</div>;
  }

  return (
    <div>
      <div>
        <h2>Reverse word</h2>
        <p>
          Hello, welcome to reverse word application. Follow the instructions
          bellow.
        </p>

        <ol>
          <li>
            You can use the input field to type words. Each letter it will be
            shown as a green card.
          </li>
          <li>To reverse a word you can press the button aside.</li>
          <li>
            Each reversed word it will be saved in a history list, e a new line
            after the green cards following the same approach of one letter per
            card but with a red background.
          </li>
          <li>You can also delete letters just clicking on it.</li>
        </ol>
      </div>
      <div>
        <InputField
          onInputFieldChange={onInputFieldChangeHandler}
          value={typedWords.recentInput}
        />
        <SendReverseWordButton
          typedWords={typedWords}
          setTypedWords={setTypedWords}
          setErrorMessage={setErrorMessage}
        />
      </div>
      <div>
        <ErrorMessage
          hideMessage={!!errorMessage}
          message={errorMessage}
          cleanErrorMessageHandler={setErrorMessage}
        />
      </div>
      <div id="current-card-inputs">
        <Card
          typedWords={typedWords.recentInput}
          onSingleLetterClickHandler={onSingleLetterCardClickHandler}
        />
      </div>
      {renderWordsHistory}
    </div>
  );
};

export default CardContainer;
