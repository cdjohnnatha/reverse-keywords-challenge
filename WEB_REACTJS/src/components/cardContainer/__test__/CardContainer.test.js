import React from 'react';
import { create } from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import axios from 'axios';

import CardContainer from '../CardContainer';
import LetterCard from '../card/letterCard/LetterCard';
import Card from '../card/Card';
import '../../../../jest.setup';

const initialInputValueAsArray = ['K', 'e', 'n', 'n', 'a'];

const event = {
  target: { value: 'Tenna' },
};

let wrapper = null;
let inputFieldComponent = null;
let cardsContainer = null;
jest.mock('axios');

describe('Test CardContainer component', () => {
  test('CardContainer snapshot renders', () => {
    const component = create(<CardContainer />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  beforeEach(() => {
    wrapper = mount(<CardContainer />);
    inputFieldComponent = wrapper.find('input');
    cardsContainer = wrapper.find('.container');
  });

  test('InputField values is the same of each single card', () => {
    expect(inputFieldComponent.prop('value')).toBe('Kenna');
    expect(cardsContainer.getElement().type).toBe('div');

    cardsContainer
      .find('div')
      .getElements()
      .forEach((singleCard, index) => {
        if (typeof singleCard.props.children === 'string') {
          expect(singleCard.props.children).toBe(
            initialInputValueAsArray[index - 1]
          );
        }
      });
  });

  test('Change first letter at InputField it will change first letter at cards', () => {
    wrapper.find('input').simulate('change', event);
    expect(wrapper.find('input').prop('value')).toBe('Tenna');
    expect(wrapper.find('.container').find('div').at(1).text()).toBe('T');
    expect(wrapper.find('.container').find('div').length).toBe(6);
  });
  test('Change first letter at InputField it will change first letter at cards', () => {
    event.target.value = 'Last';
    wrapper.find('input').simulate('change', event);
    expect(wrapper.find('input').prop('value')).toBe('Last');
    expect(wrapper.find('.container').find('div').at(1).text()).toBe('L');
    expect(wrapper.find('.container').find('div').at(4).text()).toBe('t');
    expect(wrapper.find('.container').find('div').length).toBe(5);
  });
  test('Test click button to reverse words, should create a new card with not reverse word and change the recentInput with a reversed word', async () => {
    expect(wrapper.find('button').text()).toBe('Reverse the input text "Kenna"');

    let cardHistoryWrapper = wrapper.find('#card-history').find(Card);
    expect(cardHistoryWrapper.length).toBe(0);

    await act(async () => {
      await axios.get.mockResolvedValue({ data: { reverse_word: 'anneK' } });
      wrapper.find('button').simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('input').prop('value')).toBe('Kenna');
    const currentCardInputs = wrapper
      .find('#current-card-inputs')
      .find(LetterCard);

    expect(currentCardInputs.at(0).text()).toBe('K');
    expect(currentCardInputs.at(1).text()).toBe('e');
    expect(currentCardInputs.at(2).text()).toBe('n');
    expect(currentCardInputs.at(3).text()).toBe('n');
    expect(currentCardInputs.at(4).text()).toBe('a');

    cardHistoryWrapper = wrapper.find('#card-history').find(Card);
    const historyCards = wrapper.find('#card-history').find(LetterCard);
    expect(historyCards.at(0).text()).toBe('a');
    expect(historyCards.at(1).text()).toBe('n');
    expect(historyCards.at(2).text()).toBe('n');
    expect(historyCards.at(3).text()).toBe('e');
    expect(historyCards.at(4).text()).toBe('K');
    expect(cardHistoryWrapper.length).toBe(1);
  });

  test('Test click button to reverse words, should create a new history card with red-bg the recentInput should keep green', async () => {
    expect(wrapper.find('button').text()).toBe('Reverse the input text "Kenna"');

    const cardHistoryWrapper = wrapper.find('#card-history').find(Card);
    expect(cardHistoryWrapper.length).toBe(0);

    await act(async () => {
      await axios.get.mockResolvedValue({ data: { reverse_word: 'anneK' } });
      wrapper.find('button').simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('input').prop('value')).toBe('Kenna');
    const currentCardInputs = wrapper.find('.green-bg');
    expect(currentCardInputs.length).toBe(5);
    expect(currentCardInputs.at(0).text()).toBe('K');
    expect(currentCardInputs.at(1).text()).toBe('e');
    expect(currentCardInputs.at(2).text()).toBe('n');
    expect(currentCardInputs.at(3).text()).toBe('n');
    expect(currentCardInputs.at(4).text()).toBe('a');
    const historyCards = wrapper.find('.red-bg');
    expect(historyCards.length).toBe(5);
    expect(historyCards.at(0).text()).toBe('a');
    expect(historyCards.at(1).text()).toBe('n');
    expect(historyCards.at(2).text()).toBe('n');
    expect(historyCards.at(3).text()).toBe('e');
    expect(historyCards.at(4).text()).toBe('K');
  });
});
