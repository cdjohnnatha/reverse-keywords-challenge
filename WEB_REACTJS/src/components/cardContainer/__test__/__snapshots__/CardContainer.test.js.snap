// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CardContainer component CardContainer snapshot renders 1`] = `
<div>
  <div>
    <h2>
      Reverse word
    </h2>
    <p>
      Hello, welcome to reverse word application. Follow the instructions bellow.
    </p>
    <ol>
      <li>
        You can use the input field to type words. Each letter it will be shown as a green card.
      </li>
      <li>
        To reverse a word you can press the button aside.
      </li>
      <li>
        Each reversed word it will be saved in a history list, e a new line after the green cards following the same approach of one letter per card but with a red background.
      </li>
      <li>
        You can also delete letters just clicking on it.
      </li>
    </ol>
  </div>
  <div>
    <input
      className="field input-style"
      onChange={[Function]}
      style={Object {}}
      type="text"
      value="Kenna"
    />
    <button
      className="button-style"
      onClick={[Function]}
      type="button"
    >
      Reverse the input text "
      Kenna
      "
    </button>
  </div>
  <div />
  <div
    id="current-card-inputs"
  >
    <div
      className="container"
    >
      <div
        className="card-class flex-container box-margin green-bg"
        onClick={[Function]}
      >
        K
      </div>
      <div
        className="card-class flex-container box-margin green-bg"
        onClick={[Function]}
      >
        e
      </div>
      <div
        className="card-class flex-container box-margin green-bg"
        onClick={[Function]}
      >
        n
      </div>
      <div
        className="card-class flex-container box-margin green-bg"
        onClick={[Function]}
      >
        n
      </div>
      <div
        className="card-class flex-container box-margin green-bg"
        onClick={[Function]}
      >
        a
      </div>
    </div>
  </div>
</div>
`;
